src/core/analyzer-cache-context.ts

```diff
@@ -0,0 +1,483 @@
+    copy._cache = cache;
+    copy._generation = this._generation + 1;
+    return copy;
+  }
+
+  /**
+   * Scan a toplevel document given its url and optionally its contents.
+   */
+  private async _scan(
+      resolvedUrl: string, contents?: string,
+      visited?: Set<string>): Promise<ScannedDocument|'visited'> {
+    if (visited && visited.has(resolvedUrl)) {
+      return 'visited';
+    }
+    visited = visited || new Set();
+    visited.add(resolvedUrl);
+    const cachedResult = this._cache.scannedDocumentPromises.get(resolvedUrl);
+    if (cachedResult) {
+      await this._scanDependenciesOfToplevelDoc(await cachedResult, visited);
+      return cachedResult;
+    }
+    const promise = (async() => {
+      // Make sure we wait and return a Promise before doing any work, so that
+      // the Promise is cached before anything else happens.
+      await Promise.resolve();
+      const document = await this._parse(resolvedUrl, contents);
+      return this._scanDocument(document, visited);
+    })();
+    this._cache.scannedDocumentPromises.set(resolvedUrl, promise);
+    const scannedDocument = await promise;
+    await this._scanDependenciesOfToplevelDoc(scannedDocument, visited);
+    return scannedDocument;
+  }
+
+  /**
+   * Parses and scans a document from source.
+   */
+  private async _scanInlineSource(
+      type: string, contents: string, url: string, visited: Set<string>,
+      inlineInfo?: InlineDocInfo<any>,
+      attachedComment?: string): Promise<ScannedDocument> {
+    const resolvedUrl = this._resolveUrl(url);
+    const parsedDoc =
+        this._parseContents(type, contents, resolvedUrl, inlineInfo);
+    const scannedDoc =
+        await this._scanDocument(parsedDoc, visited, attachedComment);
+    return scannedDoc;
+  }
+
+  /**
+   * Scans a ParsedDocument.
+   */
+  private async _scanDocument(
+      document: ParsedDocument<any, any>, visited: Set<string>,
+      maybeAttachedComment?: string): Promise<ScannedDocument> {
+    const warnings: Warning[] = [];
+    const scannedFeatures = await this._getScannedFeatures(document);
+    // If there's an HTML comment that applies to this document then we assume
+    // that it applies to the first feature.
+    const firstScannedFeature = scannedFeatures[0];
+    if (firstScannedFeature && firstScannedFeature instanceof ScannedElement) {
+      firstScannedFeature.applyHtmlComment(maybeAttachedComment);
+    }
+
+    const scannedDocument =
+        new ScannedDocument(document, scannedFeatures, warnings);
+
+    if (!scannedDocument.isInline) {
+      if (this._cache.scannedDocuments.has(scannedDocument.url)) {
+        throw new Error(
+            'Scanned document already in cache. This should never happen.');
+      }
+      this._cache.scannedDocuments.set(scannedDocument.url, scannedDocument);
+    }
+    await this._scanInlineDocuments(scannedDocument, visited);
+
+    return scannedDocument;
+  }
+
+  private async _scanDependenciesOfToplevelDoc(
```

Comment:
+1. Added a bunch of docs and also did some mild reorganization for clarity.

The organization of code into private methods here had gotten really spread out and hard to follow, but thanks to other changes in this PR I was able to simplify through a bit of consolidation.